### ISSUE TITLE
ENH: No longer auto-convert array scalars to numpy scalars in ufuncs (and elsewhere)

### ISSUE THREAD

**Issue by mhvk** [numpy/numpy#24897]
### Proposed new feature or change:

This may come too late, but numpy 2.0 may be a good opportunity to remove an irritating wart, that `ufunc(array_scalar_arguments)` returns a numpy scalar rather than an array scalar. There is a very long-standing PR (#14489) that introduces an option to avoid this (using `out=...`), but perhaps it is time to just ditch the conversion to scalar? 

See #13015 for discussion. As a specific example of why it is unhandy, it mentions:
> #13100 raises a case where `np.fix` resorts to calling `np.asanyarray(np.ceil(x, out=out))` in order to ensure that the result is an array. Unfortunately, this has a number of draw-backs:
> 
> * It discards duck-types, like `dask` arrays
> * It changes the dtype of 0d object arrays containing array-likes

The [first comment](https://github.com/numpy/numpy/issues/13105#issuecomment-471382946) on the issue by @rgommers suggests just removing the cast to scalar. To me, this still makes the most sense.

**Comment by seberg** [comment]
**Reposting from gh-14489, not everything aged perfectly:**

An alternative thing that would be a big change, but not as big as removing scalars one, would be this: https://github.com/seberg/numpy/pull/new/try-0d-preservation that I tried to explore a bit today.

The proposal would be to:
* Ensure that 0-d arrays are preserved in ufuncs, `np.add(0d_array, 2)` will return a 0d array.  (And all other functions, this only has ufuncs right now!)
* Reductions would return scalars if `axis=None`.
* We may or may not need a (internal) helper `arr, wrap = asarray_and_wrap(obj)` that returns a function to do the "decaying".

(The tail of that is not super trivial, quite a few test failures, scipy also relies on scalars in a few places for example, but it looks harmless).  But run that branch with:
```
NPY_NUMPY_2_BEHAVIOR=1
```
to try.  A fun little caveat: `np.matmul(vector, vector)` should be a scalar so is a bit hackish still using the old behavior due to that.  

---

I never truly believed that we can pull of *removing* scalars or always converting them to arrays in ufuncs.  But maybe I am just too far on the side of "change as little as possible".

The alternative is what I say above.  That means making basically scalars something like NumPy duck-types with an implicit `__array_wrap__` that unwraps them, but they have a lower array-priority (one way to make a lot of our Python code be able to retain scalars when they should).

In general, am happy with any such thoughts, but really need to focus a bit more on NEP 50 (the overlap isn't that large).
The note above with `arr, wrap = asarray_and_wrap(obj)` has some overlap, because NEP 50 has the some use in getting the result-type in a python scalar aware way more conveniently.

---

In the end, NumPy 2.0 clearly will slip a while.  But one thing I want to note is that while I would love this to be pushed, I also still think that NEP 50 and a few C-API cleanups should be the main must-have blockers.
(Not that I am against more, but if it turns out they come with tricky downstream fallout, that would make me ask the question:  Isn't it better to do what we must do now, and just wait a few years, as painful as it is?)

**Comment by mdhaber** [comment]
@h-vetinari thought you would be interested in this one.

**Comment by georgeef** [comment]
Not in the scope of this discussion, but anyway related: from the user's point of view, a 0-d ndarray with an empty shape is mostly a source of confusion. A scalar can be better represented as a 1-d ndarray with shape (1,) and size 1, if needed. It is fine to have (semi-)automatic conversions from a size 1 array to a scalar, with all the confusion that this creates (as with 0-d arrays). But I don't see the need to have a third representation for the same mathematical concept. What is the gap filled by 0-d arrays?

An empty list is a list which has nothing inside. It is different than any non-empty list, so it is necessary, in order to fill this gap. A 0-d array with an empty shape has something inside. It is just another representation for something already available. A completely empty ndarray (with size 0) wouldn't make sense, and it would violate the rule that the total size is the product of the axes sizes.

With respect to broadcasting, a 0-d array expands to the shape of the others. An 1-d array of size 1 does the same, also for trivial broadcastings.

With respect to shape concatenation, e.g., outer product, all representations (scalar, 0-d array, 1-d array of size 1) add a new axis of size 1, so this operation is consistent for 1-d arrays and irregular for 0-d arrays (there should be no new axis after concatenating an empty shape).

Perhaps there are edge cases where a 0-d array makes more sense, but I think the inconsistencies that it creates are more than the benefit.

**Comment by jorenham** [comment]
> Not in the scope of this discussion, but anyway related: from the user's point of view, a 0-d ndarray with an empty shape is mostly a source of confusion.

I remember that I have been confused by this in the past, because I expected a scalar, but instead got something that was *almost* a scalar, but slightly different. But if there wouldn't have been scalars at all, then I also wouldn't have been confused about it.

**Comment by georgeef** [comment]
Yes, this is one approach, I would say it is the theoretically correct approach. You represent scalars as 0-d arrays, and what are scalar now become syntactic sugar to create 0-d arrays. Internally all scalars are converted to 0-d arrays, or at least they behave as if they had been implemented this way.

However ...

Scalars are probably the most useful objects, and for most practical purposes it makes sense to have dedicated representation for them. Then you ditch the 0-d arrays, because you already have them. So the second approach is to forbid 0-d arrays and to exchange scalars with 1-d arrays of size 1 when needed. With this approach all scalars behave as if they were 1-d arrays of size 1, but they have distinct implementation.

From theoretical perspective, the second approach is not so nice (scalars in mathematics are 0-d objects, not vectors with one coordinate). In most cases the two approaches are equivalent, in corner cases they may differ. The important question is: when they differ, which result do most users expect? And which approach is easier to implement and to document? E.g., having a rule with a dedicated branch for the case of 0-d arrays is disaster for comprehension and documentation. Changing a phase like "where idx is a tuple of indexes" to "where idx is a non-empty tuple of indexes" is an inconvenience but not disaster. I cannot answer this question (I have no idea about the implementation and also little understanding about what users want).

An analogy: you want to fill a room and you have a nice box which fits perfectly to the space. However, in one corner there is a heavy object since long time, which no-one wants to move. So you cut one corner of the nice box and place it in the space. Together with the separate object they perfectly cover the space.

**Comment by seberg** [comment]
Sorry, but to me this is a very weird theoretical discussion and yes, I think not having 0-D arrays would be far far worse then even the current state.  Sure, 0-D arrays can confuse you, but you replace confusion with inconsistency.
And no, 0-D arrays are *not* a hassle to implement/have around, there are some *very minor* checks in C-code needed for them, but that's about it.
(Scalars are a hassle, of course.  But having scalars and no 0-D array would be utterly terrible implementation wise.  There is an old write-up here https://numpy.org/neps/nep-0027-zero-rank-arrarys.html, I am not sure what is all argued there, a lot is probably outdated, but...)

(Another point, users rarely have 0-D arrays to begin with, unless they have code dealing with arbitrary dimensions, and while most code wouldn't notice, you would just make N-D be a weird: N, but not 0-D.)

The hassle is **purely** the fact that 0-D arrays are not type-safe.  Scalars should promote to arrays where needed, but NumPy, for historical reasons maybe, thinks it's a good idea to demote 0-D arrays to scalars.  And that demotion is very lossy and leads to issues.

I would prefer if we can keep the issue focused on that, although the newer discussion in scientific-python or the PR may replace it anyway soon.

**Comment by georgeef** [comment]
Thank you for the link. I knew one difference, I learned 1-2 more, which makes 0-d arrays even more confusing.

Sorry for the divergence, I think the fact that there are two representations for (almost) the same thing is at the core of this issue.

Back to the main topic: if you don't demote, you give a 0-d array to the users, which in most cases is not what they want (as you say, "users rarely have 0-D arrays to begin with"). If you always demote, effectively you make 0-d arrays a hidden feature (available but difficult to use). If you demote by default sometimes and not other times, it is confusing because there is no simple rule.

My preference is to demote by default and to provide an option to skip the last conversion.

**Comment by seberg** [comment]
> My preference is to demote by default and to provide an option to skip the last conversion.

Well, thats why we added `out=...` to ufuncs.  But it is very NumPy specific, so it doesn't really solve much and is hard to adopt (and also very hard to know about). It's useful inside NumPy, but I wouldn't be surprised if almost nobody uses it outside of NumPy.

So unfortunately, I doubt it would actually solve what annoys e.g. SciPy.

> If you always demote, effectively you make 0-d arrays a hidden feature (available but difficult to use).

That's exactly what we have, and it's a problem because you need them often enough.

> If you demote by default sometimes and not other times, it is confusing because there is no simple rule.

There _are_ (relatively) simple rules.  Also the one important rule would be `array + array -> array`, which should obvious?

> Back to the main topic: if you don't demote, you give a 0-d array to the users, which in most cases is not what they want (as you say, "users rarely have 0-D arrays to begin with")

IMO, the problem here really isn't (directly) changing the NumPy rules, though.  Users don't do `scalar = np.asarray(1)` normally, they get scalars via `arr.sum()` or `arr1d[0]` and *I* don't want to change that.
(Others often disagree, mainly on grounds of "no scalars at all". Which what you quote is about: I just don't really think its feasible and the reason for that is that scalars have use-cases and we can't just break them... if you started from scratch those use-cases might have been served well enough with other solution -- or explicit "to scalar" conversion maybe).

Users wouldn't get 0-D arrays randomly, but rather in contexts where they likely need them (i.e. working with N-D arrays).

The problem with changing this is really more about indirectly changing _downstream_ library code (and users also have some "library like" code), rather than more end-user code, I think.
Libraries (like SciPy) contains functions such as:
```
def func(obj):
    obj = np.asarray(obj)
    return -obj  # scipy often ends with obj[()] to demote to scalar, though.
```
and changing NumPy would change it's behavior to return always arrays due to the `asarray()`.  
So such a function was before "like a numpy ufunc" and afterwards would be a function that always returns arrays.  Which is fine sometimes, but sometimes may be confusing.
(Typing is always saner, but you may not guess from the function what kind of typing you get...)

In NumPy, we can fix this quite easily, but I am not sure if downstream libraries want to adopt a pattern of:
```
def func(obj):
    conv = array_converter(obj)
    obj, = conv

    return conv.wrap(obj)
```
(unless we can find a better one).
I also think it is actually not so bad if parts of downstream say: these functions work on arrays and never return scalars.

TBH, overall, I guess I don't think that 0-D arrays are _that_ confusing.  Yes, you are confused by a shape of `()`, etc.
But it seems a bit like a one off similar to learning the difference between mutable and immutable types: it really bites you once or a few times, but I don't think it's something that'll confuse you again and again.

(And in a perfect world _with_ scalars, I think scalars shouldn't pretend to be containers/arrays. Unfortunately, they have to, though...  It's all about feasible improvements in the end.)


EDIT: Anyway, I am out of this discussion. I think I have said about anything I can and it doesn't help to repeat it.  There needs to be a clear proposal and then the community can decide.  And neither of our opinions will matter exactly anyway.

**Comment by georgeef** [comment]
Yes, it is out of the discussion and maybe a closed topic, but it is a pain point for users like me (not library programmers).

I can imagine that in operations which naturally generate a 0-d array, it is much simpler to return it as it is. No extra code and run time for conversion, no extra cooking of data (which needs documentation), and it fits well to abstract mathematical concepts.

This is how I see it from the user's point of view:
- Dealing with two scalar representations is inconvenient and (to me) more inconsistent than having an exception for a very special case in general dimensionality rules.
- If I had to stick to one of the two representations, either numpy scalars or 0-d arrays, I would definitely choose the first. I use scalars too often and I don't think of them as 0-d arrays. I wouldn't create a 0-d array if a numpy scalar can do the same job.
- Scalars are immutable. Fine. If I need something like a buffer with a single element inside it, I would implement it as a vector (1-d array) with size 1, rather than a scalar.
- (This is one of the arguments for 0-d arrays in the link above) If I need a view to shared data, I would think of it as a vector (1-d array), even if its size is 1, rather than a scalar.
- The extra conversions to demote 0-d arrays to scalars should never be a performance issue. If scalar conversions become a bottleneck, I should better re-think my solution, or which library is good for the problem (numpy is not for massive scalar processing).

I had to learn about 0-d arrays because they pop-up where I don't need them. It would be better if users like me don't need to know anything about 0-d arrays, unless they explicitly create them. Which means numpy should accept 0-d arrays as input, but should not generate them by default as output.

**Comment by georgeef** [comment]
> thats why we added out=... to ufuncs

It is not obvious at all how this is related to scalar demotion. An argument like `demote_scalar=True` would be easier to use (of course it is not applicable if `out=` is also given).

**Comment by h-vetinari** [comment]
> from the user's point of view, a 0-d ndarray with an empty shape is mostly a source of confusion. A scalar can be better represented as a 1-d ndarray with shape (1,) and size 1, if needed.

That's absolutely not the case if you have anything that's supposed work generically across dimensions. For example, summing across a single dimension generally takes some N dimensional input, and yields an N-1 dimensional output. A lot of computations depend on the composability of these operations, and thus proper dimensionality. If suddenly the sum across a one-dimensional array were again one-dimensional (according to your "scalar can be better represented as 1-d"), _a lot_ would break.

**Comment by georgeef** [comment]
It would be a different rule, but it would still be a consistent rule. Currently the following is allowed:
- `sum(zero_d_array, axis=0)` -> returns scalar (equivalent to 0-d array)
- `sum(scalar, axis=0)` -> returns scalar

So the complete rule is: `sum` reduces `ndim` by 1; unless it is 0, then it stays 0 because it cannot go lower.

The new rule would be similar: `sum` reduces `ndim` by 1; unless it is 1, then it stays 1 because it cannot go lower.

No doubt that any change in basic rules is gong to break the existing implementation and dependencies. What I'm saying is that the current rules are not very strict either. As commented above, `outer` is another operation which doesn't follow strict rules for corner cases.
